### PR TITLE
Passthrough AD in ode with complex numbers

### PR DESCRIPTION
**Pull request by AmitRotem**
see issue https://github.com/SciML/DiffEqBase.jl/issues/858

### CODE REVIEW

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by AmitRotem** [comment]
Seems that `real(eltype(u0))` used for the complex case fails on other types, like unitful quantity.
Specialized `promote_u0` for `u0::AbstractArray{Complex}`

File: test/downstream/complex_number_ad.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+function loss_via_real(p)
+  prob = remake(prob0_real; p)
+  sol = solve(prob)
+  u = [complex.(selectdim(u,3,1), selectdim(u,3,2)) for u=sol.u]
+  cost(u) + sum(p) / 10
+end
+
+# assert
+@assert eltype(last(solve(prob0     ).u)) <: Complex
+@assert eltype(last(solve(prob0_real).u)) <: Real
```
Comment (ChrisRackauckas) [inline test/downstream/complex_number_ad.jl:56]:
```suggestion
  sol = solve(prob, Tsit5())
  cost(sol.u) + sum(p) / 10
end

## same problem via reals
### realify complex ode problem
function real_f(du,u,p,t)
  complex_u = complex.(selectdim(u,3,1), selectdim(u,3,2))
  complex_du = copy(complex_u)
  prob0.f(complex_du, complex_u, p, t)
  selectdim(du,3,1) .= real(complex_du)
  selectdim(du,3,2) .= imag(complex_du)
  return nothing
end
prob0_real = remake(prob0; f=real_f, u0=cat(real(prob0.u0), imag(prob0.u0); dims=3))
### real loss function via real ode
function loss_via_real(p)
  prob = remake(prob0_real; p)
  sol = solve(prob, Tsit5())
  u = [complex.(selectdim(u,3,1), selectdim(u,3,2)) for u=sol.u]
  cost(u) + sum(p) / 10
end

# assert
@assert eltype(last(solve(prob0, Tsit5()     ).u)) <: Complex
@assert eltype(last(solve(prob0_real, Tsit5()).u)) <: Real
```

**Comment by AmitRotem** [comment]
This should solve the complex ad test.
Not sure what is the problem with the `Sundails.jl` test.
Nor the format test.

**Comment by ChrisRackauckas** [comment]
The Sundials thing is unrelated. The formatting failure is because it's not following the standard formatting rules https://docs.sciml.ai/SciMLStyle/stable/, but I'll just run the formatter after merging. In the future, remember to do:

```julia
using JuliaFormatter, SomePackage
format(joinpath(dirname(pathof(SomePackage)), ".."))
```

**Comment by AmitRotem** [comment]
Found a small issue with complex numbers. The Dual of time in the ode function has NaN values. Meaning, in the ode function `(u,p,t) -> ...` `t` is `Dual` with derivative `NaN`
Maybe a problem with the promotion of `tspan` ? affecting the initial `dt` value ?

Example;
```
using LinearAlgebra, DifferentialEquations
import ForwardDiff as FD

H0 = randn(2,2)
u0 = [1.0, 0im]
Ht(u,p,t) = (H0*u)*cos(p*t)
prob0 = ODEProblem(Ht, u0, (0.0,1.0))

function loss(p)
    prob = remake(prob0; p)
    sol = solve(prob)
    lo = abs2(tr(first(sol.u)'last(sol.u)))
    lo
end

loss(rand())
FD.derivative(loss, rand()) # NaN's
```

This does not show up in other cases, such as;
- if the initial difference is small; e.g., `Ht(u,p,t) = (H0*u)*sin(p*t)`
- other random initial state; `u0 = randn(ComplexF64,2)`
- real initial state `u0 = [1.0, 0.0]`
- use some finite `dt` in `ODEProblem(...; dt=1)`

**Comment by ChrisRackauckas** [comment]
Open a new issue for this.

**Pull request merged**